File: leetcode2/2medium/최원준/Q1218.java
package Leetcode.최원준;

/*
1. 아이디어 :
해시맵을 사용하여 각 숫자에 대해 이전 숫자와의 차이가 difference인 경우의 길이를 저장

2. 시간복잡도 :
O( n )

3. 자료구조/알고리즘 :
해시맵 / dp
 */

import java.util.HashMap;
import java.util.Map;

public class Q1218 {
    class Solution {
        public int longestSubsequence(int[] arr, int difference) {
            int n = arr.length;
            int ans = 1;
            int prev = arr[0];
            Map<Integer, Integer> lengths = new HashMap<>();

            for (int num : arr) {
                if (lengths.containsKey(num - difference)) {
                    int length = lengths.get(num - difference);
                    lengths.remove(num-difference);
                    if (lengths.containsKey(num)) {
                        lengths.put(num, Math.max(lengths.get(num), length+1));
                    } else {
                        lengths.put(num, length+1);
                    }
                } else {
                    lengths.putIfAbsent(num, 1);
                }
            }
            System.out.println(lengths);
            for (int length : lengths.values()) {
                ans = Math.max(ans, length);
            }
            return ans;
        }
    }
}

File: leetcode2/2medium/최은서/Q1218.js
/**
 * @param {number[]} arr
 * @param {number} difference
 * @return {number}
 */
var longestSubsequence = function (arr, difference) {
  const dict = {}; //dict[x] : x로 끝나는 부분 수열의 최고 길이
  let maximum = 0; //가장 긴 부분 수열의 길이
  for (let i = 0; i < arr.length; i++) {
    const x = arr[i];
    const prevLen = dict[x - difference] ?? 0;
    const currLen = Math.max(dict[x] ?? 0, prevLen + 1);
    //이전 값을 유지하거나 이어붙인 길이 중에 더 큰 값
    dict[x] = currLen; //x로 끝나는 최장 길이 갱신
    if (currLen > maximum) maximum = currLen; //전체 최댓값 갱신
  }
  return maximum;
};
